### PR TITLE
Feature/stream store compression

### PR DESCRIPTION
**Pull request by sandorw**
---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1357)

This PR appears much larger than it actually is because it includes commits from https://github.com/palantir/atlasdb/pull/1347 and https://github.com/palantir/atlasdb/pull/1341. All the new code is in the last commit. Hopefully the other two merge tomorrow and I can drop those changes and rebase on develop. I've applied the do not merge label since the commit includes code that will be merging in other PRs.

I'm not entirely happy with the messiness of StreamTest.java - a larger scale refactor may be in order, but it's probably outside the scope of this commit. Similarly, I've had to extend a lot of previously private methods in the stream store code, indicating that maybe a refactor is needed there as well. I defer to the atlasdb maintainers there.

The generated java files checked in represent the files rendered by StreamStoreRenderer.java.

### CODE REVIEW

**Comment by sandorw** [comment]
Force pushed to rebase since https://github.com/palantir/atlasdb/pull/1347 merged.

**Comment by gsheasby** [comment]
Aiming to pick this up in the week after winter break.

**Comment by sandorw** [comment]
Sounds good - will rebase sometime before everyone gets back.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java [hunk 1]
```diff
@@ -445,6 +487,28 @@ public void startSecondAndFinish(Transaction t, long streamId) {
         assertNotNull(getStream(streamId));
     }
 
+    @Test
+    public void testStreamCompression() throws IOException {
+        int inputBlocks = 4;
+        byte[] input = getCompressibleBytes(inputBlocks * StreamTestWithHashStreamStore.BLOCK_SIZE_IN_BYTES);
+
+        long id = storeStream(compressedStore, input, PtBytes.toBytes("ref"));
+        StreamMetadata metadata = txManager.runTaskReadOnly(t -> {
```
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: This lambda's pretty long - it might look better in its own method.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java [hunk 2]
```diff
@@ -445,6 +487,28 @@ public void startSecondAndFinish(Transaction t, long streamId) {
...
+            StreamTestWithHashStreamMetadataTable table = StreamTestTableFactory.of()
+                    .getStreamTestWithHashStreamMetadataTable(t);
+            StreamTestWithHashStreamMetadataRow row = StreamTestWithHashStreamMetadataRow.of(id);
+            return table.getRow(row).get().getMetadata();
+        });
+        long numBlocksUsed = getStreamBlockSize(metadata);
+
+        assertTrue(numBlocksUsed < inputBlocks);
```
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: Can we make a stronger assertion here?

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java [hunk 3]
```diff
@@ -516,6 +536,95 @@ private void unmarkStreamsAsUsed() {
                     line("index.delete(toDelete);");
                 } line("}");
             }
+
+            private void storeStreamWithCompression() {
+                line("@Override");
+                line("public Pair<", StreamId, ", Sha256Hash> storeStream(InputStream stream) {"); {
+                    line("long id = storeEmptyMetadata();");
+                    line();
+                    line("MessageDigest digest = Sha256Hash.getMessageDigest();");
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: Duplicated (generation) code here. Consider extracting the repeated group of `line()`s, or better yet, generate a private helper method.

File: docs/source/release_notes/release-notes.rst [hunk 4]
```diff
@@ -61,6 +61,12 @@ develop
          - We now publish a runnable distribution of AtlasCli that is available for download directly from Bintray
            (`Pull Request <https://github.com/palantir/atlasdb/pull/1345>`__)
 
+    *    - |new|
+         - AtlasDB now supports stream store compression.
+           Streams can be compressed client-side by adding the ``compressStreamInClient`` option to the stream
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: Should also have a docs entry exposing this possibility.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java [hunk 5]
```diff
@@ -179,6 +185,84 @@ protected void loadSingleBlockToOutputStream(Transaction t, Long streamId, long
...
+    public Map<Long, InputStream> loadStreams(Transaction t, Set<Long> ids) {
+        Map<Long, InputStream> compressedStreams = super.loadStreams(t, ids);
+        return Maps.transformValues(compressedStreams, stream -> {
+            return new LZ4BlockInputStream(stream);
+        });
+    }
+
+    @Override
+    protected void tryWriteStreamToFile(Transaction transaction, Long id, StreamMetadata metadata, FileOutputStream fos) throws IOException {
+        try (InputStream blockStream = makeStream(id, metadata);
```
Comment (gsheasby) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: Does this compile? Looks like `AbstractGenericStreamStore.makeStream` is still private.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 6]
```diff
@@ -139,7 +139,7 @@ private int getNumberOfBlocksThatFitInMemory() {
     }
 
     @Override
-    public final File loadStreamAsFile(Transaction transaction, ID id) {
+    public File loadStreamAsFile(Transaction transaction, ID id) {
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: This doesn't appear to be overridden anywhere.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java [hunk 7]
```diff
@@ -156,7 +156,7 @@ private File loadToNewTempFile(Transaction transaction, ID id, StreamMetadata me
         }
     }
 
-    private void checkStreamStored(ID id, StreamMetadata metadata) {
+    protected void checkStreamStored(ID id, StreamMetadata metadata) {
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: When you override a method in generated code, it would be good to add a comment saying that changing this method could affect the generated code in unexpected ways.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java [hunk 8]
```diff
@@ -144,10 +144,14 @@ public Sha256Hash apply(StreamMetadata metadata) {
         return hashes;
     }
 
-    protected final StreamMetadata storeBlocksAndGetFinalMetadata(@Nullable Transaction t, long id, InputStream stream) {
+    protected final StreamMetadata storeBlocksAndGetFinalMetadata(@Nullable Transaction t, long id, InputStream stream, boolean computeHash) {
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java]: ...well, I see why you're doing that now (using `false` in the generated code). Difficult to immediately see the best solution. Will try and have a look later, or discuss with you when you're online.

Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java]: It's also possible that we can just leave this out - it's just a performance optimization to avoid taking a hash that we'll just throw away. I haven't measured how much we actually save here.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/stream/StreamStoreDefinition.java [hunk 9]
```diff
@@ -32,29 +32,35 @@
...
     }
 
     public StreamStoreRenderer getRenderer(String packageName, String name) {
-        return new StreamStoreRenderer(Renderers.CamelCase(longName), idType, packageName, name, inMemoryThreshold);
+        return new StreamStoreRenderer(Renderers.CamelCase(longName), idType, packageName, name, inMemoryThreshold,
+                compressStream);
     }
 
-    public Multimap<String, Supplier<OnCleanupTask>> getCleanupTasks(String packageName, String name, StreamStoreRenderer renderer, Namespace namespace) {
+    public Multimap<String, Supplier<OnCleanupTask>> getCleanupTasks(String packageName, String name,
```
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/stream/StreamStoreDefinition.java]: Defer to you here - I didn't change anything on this method, was just cleaning up line lengths a little bit to get them <= 120. I'd prefer to have any further changes here be in another PR.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java (hunk 6 above)
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: Correct, good catch. A previous version only overrode this method and copied a bunch of other code to avoid making private methods protected. I switched to the current version (only overriding tryWriteStreamToFile) to change and duplicate as little code as possible.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java (hunk 7 above)
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractGenericStreamStore.java]: Removed this one, since it's no longer needed, but added comments in other similar cases.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java [hunk 10]
```diff
@@ -49,7 +49,7 @@ protected AbstractPersistentStreamStore(TransactionManager txManager) {
         super(txManager);
     }
 
-    private final void storeMetadataAndIndex(final long streamId, final StreamMetadata metadata){
+    protected final void storeMetadataAndIndex(final long streamId, final StreamMetadata metadata) {
```
Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java]: There are several cases like this where a now protected method is called in generated code. I can add comments for them, but I'm not sure they're needed, since the protected modifier should communicate already that changes in behavior could break others.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java (hunk 2 above)
Comment (sandorw) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/StreamTest.java]: Sure, the answer should be deterministic.

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java (hunk 5 above)
Comment (sandorw) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/schema/stream/generated/StreamTestWithHashStreamStore.java]: Fixed and will regenerate the code now. This must have been a change in your PR that I hadn't picked up.

**Comment by sandorw** [comment]
I refactored the storage/hashing code as we discussed earlier. As a bonus, the generated code that's different for the compressed case is now simpler.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java [hunk 11]
```diff
@@ -516,6 +534,54 @@ private void unmarkStreamsAsUsed() {
...
             }
+
+            private void storeBlocksAndGetFinalMetadata() {
+                line("@Override");
+                line("protected StreamMetadata storeBlocksAndGetFinalMetadata(Transaction t, long id, InputStream stream) {"); {
+                    line("//Hash the data before compressing it");
+                    line("MessageDigest digest = Sha256Hash.getMessageDigest();");
+                    line("InputStream hashingStream = new DigestInputStream(stream, digest);");
+                    line("InputStream compressingStream;");
+                    line("try {"); {
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: I think we want this to be try-with-resource for the compressingStream to ensure the streams are closed if an exception occurs during storeBlocks

File: atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java (hunk 10 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/stream/AbstractPersistentStreamStore.java]: Ack, looks ok to me.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java [hunk 12]
```diff
@@ -516,6 +534,52 @@ private void unmarkStreamsAsUsed() {
                     line("index.delete(toDelete);");
                 } line("}");
             }
+
+            private void storeBlocksAndGetFinalMetadata() {
+                line("@Override");
+                line("protected StreamMetadata storeBlocksAndGetFinalMetadata(Transaction t, long id, InputStream stream) {"); {
+                    line("//Hash the data before compressing it");
+                    line("MessageDigest digest = Sha256Hash.getMessageDigest();");
+                    line("InputStream hashingStream = new DigestInputStream(stream, digest);");
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: Do we need to wrap this line in try-with-resources, too?

Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: ideally yes, though in reality it will be closed when the `compressingStream` is closed, and a `DigestInputStream` doesn't currently do anything on close (though it could in the future).

Comment (sandorw) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/StreamStoreRenderer.java]: I'll add it in and check where it should be added in the abstract classes as well.

**Comment by sandorw** [comment]
Added try with resources to the non-overridden code as well. Force pushed to resolve the merge conflict.

**Comment by sandorw** [comment]
I believe I've addressed all the comments. Glenn, this should be good to merge whenever you sign off on it.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/ImportRenderer.java [hunk 13]
```diff
@@ -25,7 +25,7 @@
 import com.google.common.collect.Iterables;
 
 public class ImportRenderer extends Renderer {
-    private static final List<String> IMPORT_PREFIXES = ImmutableList.of("java.", "javax.", "org.", "com.");
+    private static final List<String> IMPORT_PREFIXES = ImmutableList.of("java.", "javax.", "org.", "com.", "net.");
```
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/ImportRenderer.java]: For some reason, this slightly breaks the import rendering - `renderImports` should not add a blank line when there are no imports in the block (as is the case with `net.` in most files). However, extra blank lines are added.

Running `./gradlew generateSchemas` will expose this.

Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/ImportRenderer.java]: Tracked it down: #1413.
Will merge develop into this branch, then fix a minor indentation issue, then generate schemas, then merge.
